### PR TITLE
Precompile method before running them on worker threads.

### PR DESCRIPTION
**Pull request by yuyichao**
Just clearing a long time todo in the threading code, unlikely to have any effect in simple bencharks.

### CODE REVIEW

**Comment by JeffBezanson** [comment]
IIUC, we were already pre-compiling the method; this just lifts the lookup?

**Comment by yuyichao** [comment]
You are right, it seems that we just constructs the tuple type and do the lookup before throwing it away? That part can be removed now?

File: src/gf.c [hunk 1]
```diff
@@ -2209,6 +2194,18 @@ JL_DLLEXPORT jl_value_t *jl_apply_generic(jl_value_t **args, uint32_t nargs)
...
+
+jl_method_instance_t *jl_lookup_generic(jl_value_t **args, uint32_t nargs, uint32_t callsite)
+{
+    return jl_lookup_generic_(args, nargs, callsite);
+}
+
+JL_DLLEXPORT jl_value_t *jl_apply_generic(jl_value_t **args, uint32_t nargs)
+{
+    jl_method_instance_t *mfunc = jl_lookup_generic(args, nargs,
+                                                    jl_int32hash_fast(jl_return_address()));
```
Comment (vtjnash) [inline src/gf.c]: Call the _ method?

File: src/gf.c [hunk 2]
```diff
@@ -2209,6 +2194,18 @@ JL_DLLEXPORT jl_value_t *jl_apply_generic(jl_value_t **args, uint32_t nargs)
     if (traceen)
         jl_printf(JL_STDOUT, " at %s:%d\n", jl_symbol_name(mfunc->def->file), mfunc->def->line);
 #endif
+    return mfunc;
+}
+
+jl_method_instance_t *jl_lookup_generic(jl_value_t **args, uint32_t nargs, uint32_t callsite)
```
Comment (vtjnash) [inline src/gf.c]: I think this method should take world age as a parameter, to make its presence explicit

**Review by vtjnash — APPROVED** [review]

File: src/gf.c (hunk 1 above)
Comment (yuyichao) [inline src/gf.c]: ?

Comment (yuyichao) [inline src/gf.c]: Ah, yeah, I see what you mean. Just noticed that too....

File: src/gf.c (hunk 2 above)
Comment (yuyichao) [inline src/gf.c]: Sure.

**Comment by yuyichao** [comment]
Comments addressed. I also switched it to a builtin instead of a `ccall`, which removes the allocation of the svec. This should also make it easier to do inference, precompilation and dispatch at type inference time if we need to.

**Comment by JeffBezanson** [comment]
This does not need to be a builtin --- builtins are only needed if the function must be a first-class julia function. I don't see why a builtin would be needed to avoid the svec.

**Comment by yuyichao** [comment]
What's another way of passing in variable number of arguments without allocating? For the usage in base, simply removing the `svec` seems fine, though a quick search shows that @kpamnany is using this on with multiple arguments in packages.

**Comment by kpamnany** [comment]
I'm using direct calls to the threading interface function only because currently we only support parallel loops through the Julia interface and there are a number of use cases where this is insufficient. Generally speaking, it's a terrible idea to use this interface call directly. Once `spawn()` and `sync()` are added, and `parfor()` functionality is moved into the runtime, then there will be no inappropriate exposed interfaces that can break composability.

As for whether `spawn()`, `sync()`, and `parfor()` should be builtins or not, I'd guess that they should, but don't know all the implications.

**Comment by JeffBezanson** [comment]
No they don't need to be builtins. Almost nothing needs to be a builtin. Julia functions that use `ccall` should be sufficient for those.

**Comment by yuyichao** [comment]
So how do you propose to pass a variable number of arguments to ccall without allocating the buffer?

**Comment by JeffBezanson** [comment]
I think we can ccall a varargs C function with `(Any...,)`. But, since we only ever pass one argument in Base I didn't think it was necessary. Also, in general we will be allocating a closure here, so we have to be ok with this allocation. In fact it would be simpler to require a 0-argument function, and always pass a closure.

**Comment by yuyichao** [comment]
> I think we can ccall a varargs C function with `(Any...,)`.

Nah, I think that'll be more complicated than using builtin, admittedly only in one file rather than 3...

> so we have to be ok with this allocation

Sure, though that closure is generally cheaper and smaller than the svec for simple loops. For big loops neither matters of course.

> it would be simpler to require a 0-argument function

I actually started with that but didn't want to break @kpamnany's code. I can certainly go with that and let him change the code [here](https://github.com/kpamnany/Dtree.jl/blob/4fa3ab8c4b533da470d72102dddc75a3199cf851/test/dtreebench.jl#L117-L118) to

```julia
rni = Ref(ni)
rci = Ref(ci)
rli = Ref(li)
tree = runtree(dt)
ccall(:jl_threading_run, Any, (Any,), ()->threadfun, dt, rni, rci, rli, ilock, tree, dura)
```

> No they don't need to be builtins. Almost nothing needs to be a builtin.

In the long run though, if any of them needs be statically dispatched/inferred, those should still be builtins, or we should add a path in `inference` to match those `ccall`s, which we don't have any right now.

**Comment by JeffBezanson** [comment]
Isn't `ccall` already statically dispatched? But ok, I can imagine making spawn, sync, and parfor builtins when they are official parts of the language. However builtins still have certain problems, like an inefficient calling convention. It would be nice to fix that and somehow make them less ad-hoc.

**Comment by yuyichao** [comment]
> Isn't ccall already statically dispatched?

The `ccall` is but the callback they are calling isn't.

> However builtins still have certain problems, like an inefficient calling convention.

Not too inefficient, at least not more inefficient than `ccall` with `(Any...)` or `(Any,)`.

**Comment by JeffBezanson** [comment]
Isn't it worse than `(Any,)`, since 3 arguments need to be passed?

File: src/gf.c [hunk 3]
```diff
@@ -2209,6 +2193,20 @@ JL_DLLEXPORT jl_value_t *jl_apply_generic(jl_value_t **args, uint32_t nargs)
...
+
+jl_method_instance_t *jl_lookup_generic(jl_value_t **args, uint32_t nargs, uint32_t callsite,
+                                        size_t world)
+{
+    return jl_lookup_generic_(args, nargs, callsite, world);
+}
+
+JL_DLLEXPORT jl_value_t *jl_apply_generic(jl_value_t **args, uint32_t nargs)
+{
+    jl_method_instance_t *mfunc = jl_lookup_generic_(args, nargs,
```
Comment (JeffBezanson) [inline src/gf.c]: Does `jl_lookup_generic_` actually get inlined? It's pretty big, and used in two places, so I wonder.

**Comment by yuyichao** [comment]
> since 3 arguments need to be passed?

..... Sure, I was more talking about boxing. The two constant arguments are pretty cheap.

> Does `jl_lookup_generic_` actually get inlined?

Yes. At least for me locally. It's also inlined even when I was using the no `_` version which is why I didn't notice the mistake when I checked the assembly....

**Comment by yuyichao** [comment]
Switched back to `ccall` and used a hack to not break @kpamnany 's code while avoiding the allocation in base.

**Comment by JeffBezanson** [comment]
Thank you.

I think by using `lookup_generic` instead of `compile_hint` we might be lifting less work --- the lookup doesn't compile anything, `call_method_internal` triggers compilation.

**Comment by yuyichao** [comment]
Should be fixed. I'm assuming that if a threaded loop is running in codegen/type inference, it'll also be in this state when dispatched to worker threads so I hoisted the `fptr` generation to the master thread too. The worker thread should only do API dispatch now....

**Comment by tkelman** [comment]
any concern about ParallelAccelerator or others using these internals such that we shouldn't break them during feature freeze?

**Comment by yuyichao** [comment]
https://github.com/JuliaLang/julia/pull/20854#issuecomment-284159128

**Comment by yuyichao** [comment]
None of the versions here was breaking.

**Comment by tkelman** [comment]
it would be if anyone's relying on the layout of ti_threadwork_t

**Comment by yuyichao** [comment]
No it's not even exported or accessible from anywhere.

File: src/julia.h [hunk 4]
```diff
@@ -54,6 +54,46 @@
...
+static inline T
+jl_assume(T v)
+{
+    __assume(!!v);
+    return v;
+}
+#else
+#define jl_assume(cond) (cond)
+#endif
+
```
Comment (JeffBezanson) [inline src/julia.h]: Doesn't block merging this, but maybe these definitions should be in julia_internal.h?

Comment (yuyichao) [inline src/julia.h]: Either works for me, I'll move it once I'm back to my computer

File: src/julia_internal.h [hunk 5]
```diff
@@ -208,52 +225,72 @@ jl_method_t *jl_new_method(jl_code_info_t *definition,
...
-    if (fptr.jlcall_api == 1)
-        return fptr.fptr1(args[0], &args[1], nargs-1);
-    else if (fptr.jlcall_api == 3)
-        return fptr.fptr3(meth->sparam_vals, args[0], &args[1], nargs-1);
-    else if (fptr.jlcall_api == 4)
-        return fptr.fptr4(meth, &args[0], nargs, meth->sparam_vals);
+    return NULL;
+}
+
+STATIC_INLINE jl_value_t *jl_call_fptr_internal(const jl_generic_fptr_t *fptr,
```
Comment (JeffBezanson) [inline src/julia_internal.h]: Is this fptr always the same as the one inside `meth`?

Comment (yuyichao) [inline src/julia_internal.h]: Iirc, not for the recursive compilation/inference case.

**Pull request merged**